whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,213 @@
+        if (parentList.isEmpty()){
+            return ipTree.findExact(parentResource);
+        }
+        return ipTree.findFirstMoreSpecific(IpInterval.parse(parentList.getFirst().getKey().toString()));
+    }
+
+    private IpEntry searchFirstLessSpecific(final IpTree ipTree, final IpInterval reverseIp){
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(reverseIp);
+        if (parentList.isEmpty() || !resourceExist(reverseIp, parentList.getFirst())){
+            throw new RdapException("404 Not Found", "No up level object has been found for " + reverseIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+        return parentList.getFirst();
+    }
+
+    private List<IpEntry> searchTopLevelResource(final IpTree ipTree, final IpInterval reverseIp){
+        IpEntry ipEntry;
+        try {
+            ipEntry = searchFirstLessSpecific(ipTree, reverseIp);
+        } catch (RdapException ex){
+            throw new RdapException("404 Not Found", "No top-level object has been found for " + reverseIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+
+        return List.of(loopUpLevels(ipTree, ipEntry));
+    }
+
+    private IpEntry loopUpLevels(final IpTree ipTree, IpEntry reverseIp) {
+        try {
+            reverseIp = searchFirstLessSpecific(ipTree, (IpInterval) reverseIp.getKey());
+            loopUpLevels(ipTree, reverseIp);
+        } catch (RdapException ex){
+            /*
+            * Do Nothing, end of loop
+            * */
+        }
+        return reverseIp;
+    }
+
+    private boolean resourceExist(final IpInterval ip, final IpEntry firstLessSpecific){
+        final RpslObject children = getResourceByKey(ip.toString());
+        final RpslObject rpslObject = getResourceByKey(firstLessSpecific.getKey().toString());
+        if (rpslObject == null || isAdministrativeResource(children, rpslObject)) {
+            LOGGER.debug("INET(6)NUM {} does not exist in RIPE Database ", firstLessSpecific.getKey().toString());
+            return false;
+        }
+        return true;
+    }
+
+    private static boolean isAdministrativeResource(final RpslObject children, final RpslObject rpslObject) {
+        final CIString childrenStatus = children.getValueForAttribute(AttributeType.STATUS);
+        final CIString statusAttributeValue = rpslObject.getValueForAttribute(AttributeType.STATUS);
+        return (rpslObject.getType().equals(ObjectType.INETNUM) && InetnumStatus.getStatusFor(statusAttributeValue).equals(ALLOCATED_UNSPECIFIED))
+                || (rpslObject.getType().equals(ObjectType.INET6NUM) && !Inet6numStatus.getStatusFor(childrenStatus).isAssignment() &&
+                Inet6numStatus.getStatusFor(statusAttributeValue).equals(ALLOCATED_BY_RIR));
+    }
+
+
+    @Nullable
+    private RpslObject getResourceByKey(final String key){
+        if (IpInterval.parse(key) instanceof Ipv4Resource){
+            return rpslObjectDao.getByKeyOrNull(ObjectType.INETNUM, key);
+        }
```

Comment:
parse the key and then hit the database

Reply:
I don't understand what you mean
I am already parsing and then hitting the database

Reply:
Sorry I must have looked at it wrongly